config.go

```diff
@@ -160,23 +120,74 @@ func makeCfg(config []Config) (cfg Config) {
 			cfg.AuthScheme = "Bearer"
 		}
 	}
-	if cfg.KeyRefreshTimeout == nil {
-		cfg.KeyRefreshTimeout = &defaultKeyRefreshTimeout
-	}
 
 	if cfg.KeyFunc == nil {
-		if len(cfg.KeySetURLs) > 0 {
-			jwks := &KeySet{
-				Config: &cfg,
+		if len(cfg.SigningKeys) > 0 || len(cfg.JWKSetURLs) > 0 {
+			var givenKeys map[string]keyfunc.GivenKey
+			if cfg.SigningKeys != nil {
+				givenKeys = make(map[string]keyfunc.GivenKey, len(cfg.SigningKeys))
+				for kid, key := range cfg.SigningKeys {
+					givenKeys[kid] = keyfunc.NewGivenCustom(key, keyfunc.GivenKeyOptions{}) // TODO User supplied alg?
+				}
+			}
+			if len(cfg.JWKSetURLs) > 0 {
+				var err error
+				if len(cfg.JWKSetURLs) == 1 {
+					cfg.KeyFunc, err = oneKeyfunc(cfg, givenKeys)
+				} else {
+					cfg.KeyFunc, err = multiKeyfunc(givenKeys, cfg.JWKSetURLs)
+				}
+				if err != nil {
+					panic("Failed to create keyfunc from JWK Set URL: " + err.Error()) // TODO Don't panic?
+				}
+			} else {
+				cfg.KeyFunc = keyfunc.NewGiven(givenKeys).Keyfunc
 			}
-			cfg.KeyFunc = jwks.keyFunc()
 		} else {
-			cfg.KeyFunc = jwtKeyFunc(cfg)
+			cfg.KeyFunc = func(token *jwt.Token) (interface{}, error) {
+				return cfg.SigningKey, nil
+			}
 		}
 	}
+
 	return cfg
 }
 
+func oneKeyfunc(cfg Config, givenKeys map[string]keyfunc.GivenKey) (jwt.Keyfunc, error) {
+	jwks, err := keyfunc.Get(cfg.JWKSetURLs[0], keyfuncOptions(givenKeys))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get JWK Set URL: %w", err)
+	}
+	return jwks.Keyfunc, nil
+}
+
+func multiKeyfunc(givenKeys map[string]keyfunc.GivenKey, jwkSetURLs []string) (jwt.Keyfunc, error) {
+	opts := keyfuncOptions(givenKeys)
+	multiple := make(map[string]keyfunc.Options, len(jwkSetURLs))
+	for _, url := range jwkSetURLs {
+		multiple[url] = opts
+	}
+	multiOpts := keyfunc.MultipleOptions{
+		KeySelector: keyfunc.KeySelectorFirst,
+	}
+	multi, err := keyfunc.GetMultiple(multiple, multiOpts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get multiple JWK Set URLs: %w", err)
+	}
+	return multi.Keyfunc, nil
+}
+
+func keyfuncOptions(givenKeys map[string]keyfunc.GivenKey) keyfunc.Options {
+	return keyfunc.Options{
+		// TODO Add error logger?
```

Comment:
The `keyfunc` package may log when the background goroutine fails to refresh the JWK Set. However, no logger is present in this package currently.

Is logging in this case necessary?
Should this log to the `log` package's default logger?
Should a logger be provided in the `Config` data structure?

Reply:
Is logging in this case necessary? -> yes
Should this log to the log package's default logger? -> yes

Reply:
I've added a log statement to the refresh error handler